cibuildwheel/macos.py

```diff
@@ -460,15 +461,19 @@ def build(options: Options, tmp_path: Path) -> None:
                             # rosetta2 will provide the emulation with just the arch prefix.
                             arch_prefix = ["arch", "-x86_64"]
                         else:
-                            raise RuntimeError(
-                                "don't know how to emulate {testing_arch} on {machine_arch}"
-                            )
+                            msg = f"don't know how to emulate {testing_arch} on {machine_arch}"
+                            raise RuntimeError(msg)
 
                     # define a custom 'call' function that adds the arch prefix each time
-                    def call_with_arch(*args: PathOrStr, **kwargs: Any) -> None:
-                        call(*arch_prefix, *args, **kwargs)
-
-                    def shell_with_arch(command: str, **kwargs: Any) -> None:
+                    call_with_arch = functools.partial(call, *arch_prefix)
+
```

Comment:
Personally, I'm not a massive fan of functools.partial, it's a bit high-order for my taste. I do understand what it's doing, but I really have to think about it, whereas I can just read the explicit version. What's the warning that pylint produces here?

Reply:
`functools.partial` does exactly what you expect it to do. It uses the value of arch_prefix. While cell capture (which I absolutely hate, it makes things much, much harder to read - I should not have to read the entire body of a function to see what variables it references the are not in the arguments) does _not_ do that, instead it takes the variable and captures it _later_. So this code:

```python
y = 3
def f():
    return y
y = 4
print(f"{f()=}")
```
Does not print `f()=3`, it prints `f()=4`.

This code is much more readable, doesn't require reading the body of the function, and has no surprises:

```python
y = 3
def f_x(x):
    return x
f = functools.partial(f_x, y)
y = 4
print(f"{f()=}")
```

It prints `f()=3`.

Reply:
I think these sorts of changes were:

> Cell capture in loop - this causes a value to get captured, but later - to avoid confusion or later mistakes, I've either used default value capture (which is at define site, not call site), or used functools.partial.

Reply:
"Cell capture in loop" is a valid criticism, but it doesn't apply in this case as the function isn't used outside of that iteration. I'd class this as a false positive.

Reply:
I really, really don't like capture. It's misused and makes code really hard to read.  I've even included it in a problem to students before, https://github.com/henryiii/compclass/blob/master/problems/Week10.ipynb :)

It only works in this case because arch_prefix doesn't change. However, if someone were to refactor this one day to overwrite arch_prefix later, it would suddenly change the behavior of these to functions - the dependence on `arch_prefix` is completely hidden after the fact and requires studying the contents of the functions to see.

Reply:
That's fine, but I still maintain that it's a false positive. Lint rules are friction. But they're good if they catch issues that save us as maintainers time make code review easier. This is not something that I would object to if I was reviewing the code, because it's not a bug!

Reply:
A really, really easy bug to introduce is to assign the `arch_prefix` later. The capture version is securely holding a reference to that variable, rather than capturing the current value, which is naively what most users would expect.

This makes the code much easier to read, since you don't have to inspect the contents of the function to notice it is actually capturing something in the outer scope. I'd maintain that any capture inside a function is harder to read than placing it into the signature - that's what the signature is for. I shouldn't be forced to read bodies of functions to know what variables are being used!

And yes, to the best of my knowledge, I always object to use of variable capture in reviews. (Actually, I object to nested functions in reviews, too, they reduce the readability and just tend to be used to capture variables, which is bad. I'm sure you can see such reviews in, for example, DecayLanguage & Particle where I've done exactly that. ;) )

Reply:
What about moving these nested functions out, actually! I think that would probably help quite a bit here. Yes, it did! New version is just:

```python
call_with_arch = functools.partial(call, *arch_prefix)
shell_with_arch = functools.partial(shell, *arch_prefix)
```

And no tested functions at all. :)